types/src/transaction/webauthn.rs

```diff
@@ -0,0 +1,693 @@
+use anyhow::{anyhow, Result};
+use aptos_crypto::{
+    hash::CryptoHash, p256_ecdsa, signing_message, CryptoMaterialError, HashValue, Signature,
+};
+use passkey_types::{crypto::sha256, webauthn::CollectedClientData, Bytes};
+use serde::{Deserialize, Serialize};
+
+/// Returns the binary concatenation of
+/// 1. [`authenticator_data_bytes`](PartialAuthenticatorAssertionResponse) and
+/// 2. SHA-256 hash of [`client_data_json`](PartialAuthenticatorAssertionResponse),
+///
+/// See <https://www.w3.org/TR/webauthn-3/#sctn-verifying-assertion>
+fn generate_verification_data(authenticator_data_bytes: &[u8], client_data_json: &[u8]) -> Vec<u8> {
+    // Let hash be the result of computing a hash over the clientData using SHA-256.
+    let client_data_json_hash = sha256(client_data_json);
+    // Binary concatenation of authData and hash.
+    // Note: This is compatible with signatures generated by FIDO U2F
+    // authenticators. See §6.1.2 FIDO U2F Signature Format Compatibility
+    // See <https://www.w3.org/TR/webauthn-3/#sctn-fido-u2f-sig-format-compat>
+    authenticator_data_bytes
+        .iter()
+        .chain(client_data_json_hash.iter())
+        .copied()
+        .collect::<Vec<u8>>()
+}
+
+/// This checks that the SHA3-256 of the `signing_message(message)`
+/// is equal to the actual `challenge` from `client_data_json` in `PartialAuthenticatorAssertionResponse`
+fn verify_expected_challenge_from_message_matches_actual<T: Serialize + CryptoHash>(
+    message: &T,
+    actual_challenge: &[u8],
+) -> std::result::Result<bool, CryptoMaterialError> {
+    // Generate signing_message
+    let signing_message_bytes = signing_message(message)?;
+    // Expected challenge is SHA3-256 digest of RawTransaction bytes
+    let expected_challenge = HashValue::sha3_256_of(signing_message_bytes.as_slice());
+
+    // Handle invariant where lengths are different
+    if actual_challenge.len() != expected_challenge.to_vec().len() {
+        return Err(CryptoMaterialError::WrongLengthError);
+    }
+
+    // Compute deep equal of actual challenge and message (expected challenge)
+    let deep_equal_challenge = actual_challenge
+        .iter()
+        .zip(expected_challenge.to_vec().iter())
+        .all(|(a, b)| a == b);
+
+    // Verify that message is equal to challenge
+    if deep_equal_challenge {
+        return Ok(deep_equal_challenge);
+    }
+    Err(CryptoMaterialError::ValidationError)
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum WebAuthnSignature {
+    P256Ecdsa { signature: p256_ecdsa::Signature },
+}
+
+/// `PartialAuthenticatorAssertionResponse` includes a subset of the fields returned from
+/// an [`AuthenticatorAssertionResponse`](passkey_types::webauthn::AuthenticatorAssertionResponse)
+///
+/// See <https://www.w3.org/TR/webauthn-3/#authenticatorassertionresponse>
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct PartialAuthenticatorAssertionResponse {
```

Comment:
Not a big fan of this name, but not sure what it should be.

Reply:
Closing for now. Lmk if you have a better name